contrib/opencensus-ext-azure/opencensus/ext/azure/metrics_exporter/__init__.py

```diff
@@ -12,31 +12,89 @@
 # limitations under the License.
 
-import threading
+import time
 
 from opencensus.ext.azure.common import Options
+from opencensus.metrics.export import value
+from opencensus.metrics.export.metric_descriptor import MetricDescriptorType
 from opencensus.metrics import transport
 from opencensus.stats import stats
+from opencensus.ext.azure.common import utils
+from opencensus.ext.azure.common.protocol import Data
+from opencensus.ext.azure.common.protocol import DataPoint
+from opencensus.ext.azure.common.protocol import Envelope
+from opencensus.ext.azure.common.protocol import MetricData
+from opencensus.ext.azure.common.transport import TransportMixin
 
 __all__ = ['MetricsExporter', 'new_metrics_exporter']
 
 
-class MetricsExporter(object):
+class MetricsExporter(TransportMixin):
     """Metrics exporter for Microsoft Azure Monitor."""
 
-    def __init__(self, options):
+    def __init__(self, options=None):
+        if options is None:
+            options = Options()
         self.options = options
         if not self.options.instrumentation_key:
             raise ValueError('The instrumentation_key is not provided.')
-        self._md_cache = {}
-        self._md_lock = threading.Lock()
 
     def export_metrics(self, metrics):
-        metrics = list(metrics)
+        if not metrics:
+            return
+        envelopes = []
         for metric in metrics:
-            # TODO: implement export to Azure Monitor functionality
-            print(repr(metric))
+            # No support for histogram aggregations
+            if metric.descriptor.type == MetricDescriptorType.CUMULATIVE_DISTRIBUTION:
+                continue
+            envelopes.append(self.metric_to_envelope(metric))
+            self._transmit(envelopes)
 
+    def metric_to_envelope(self, metric):
+        # The timestamp is when the metric was recorded
+        timestamp = metric.time_series[0].points[0].timestamp
+        envelope = Envelope(
+            iKey=self.options.instrumentation_key,
+            tags=dict(utils.azure_monitor_context),
+            time=timestamp.isoformat(),
+        )
+        envelope.name = 'Microsoft.ApplicationInsights.Metric'
+        data = MetricData(
+            metrics=self.metric_to_data_points(metric),
+            properties=self.get_metric_properties(metric)
+        )
+        envelope.data = Data(baseData=data, baseType="MetricData")
+        return envelope
+
+    def metric_to_data_points(self, metric):
+        """Convert an metric's OC time series to a list of Azure data points."""
+        data_points = []
+        # Each time series will be uniquely identified by it's label values
+        for time_series in metric.time_series:
+            # Using stats, time_series should only have one point
+            # which contains the aggregated value
+            for point in time_series.points:
+                if point.value is not None:
+                    data_point = DataPoint(ns=metric.descriptor.name,
```

Comment:
Would the name allow something like a namespace? E.g. https://github.com/census-instrumentation/opencensus-specs/blob/master/stats/HTTP.md#measures
If yes, we might consider to split the name string to namespace and the actual name?

Reply:
Interesting solution. Given the max length of name is 1024 characters, it is possible to encode the namespace and actual name inside the "name" field. Having a namespace field is useful as it categorizes what kind of metric its for, and the name serves as the specific type of metric. However there must be some documentation on how to use the encoding and for corner cases as well. (e.g. if we do name=<namespace>:<actual_name>, what happens when either of them is blank or the ":" is missing?)

Reply:
I'd say we leave it as it is for now, I'd rather change the DataPoint API than having to encode something in name.